ykrt/src/mt.rs

```diff
@@ -219,215 +217,206 @@ impl MTThread {
+                                // We don't want to park this thread, so even though we can see the
+                                // count is going backwards, go back to the interpreter.
+                                break;
+                            }
                         }
-                        // This Location is still being counted, so go around the loop again and
-                        // try to apply our increment.
                     }
                 }
-            }
-            PHASE_TRACING | PHASE_TRACING_LOCK => {
-                if lp.phase() == PHASE_TRACING_LOCK {
-                    // PHASE_TRACING_LOCK requires special handling, because another thread might
-                    // have grabbed it on this Location even though this thread is tracing this
-                    // location.
-
-                    if self.inner.tracer.is_none() {
-                        // If this thread isn't tracing anything, there's little point in spinning
-                        // as the thread tracing this location is still plausibly alive.
-                        return None;
-                    }
-
-                    // We wait to grab the lock.
-                    // FIXME: this is a terrible spinlock.
-                    while lp.phase() == PHASE_TRACING_LOCK {
-                        yield_now();
-                        lp = loc.load(Ordering::Acquire);
-                    }
-
-                    if lp.phase() != PHASE_TRACING {
-                        // While waiting for the Location to move beyond PHASE_TRACING_LOCK, we
-                        // probably raced with the thread tracing this location.
-                        return None;
-                    }
+                return None;
+            } else {
+                if self.inner.tracing.is_some() {
+                    // This thread is already tracing another Location, so either another
+                    // thread needs to trace this Location or this thread needs to wait
+                    // until the current round of tracing has completed. Either way,
+                    // there's no point incrementing the hot count.
+                    return None;
                 }
-
-                // Before we can do anything with this phase's value, we need to transition to
-                // `PHASE_TRACING_LOCK`.
+                // To avoid racing with another thread that may also try starting to trace this
+                // location at the same time, we need to initialise and lock the Location, which we
+                // perform in a single step. Since this is such a critical step, and since we're
+                // prepared to bail out early, there's no point in yielding: either we win the race
+                // by trying repeatedly or we give up entirely.
+                let hl_ptr = Box::into_raw(Box::new(HotLocation::Tracing(None)));
+                let new_ls = State::new().with_hotlocation(hl_ptr).with_lock();
```

Comment:
Does it make sense to have a pre-allocated initial hotlocation which we just copy into a box each time? Probably not worth the effort?

Reply:
I'm not quite sure I understand. Each location needs a pointer to a new chunk on the heap, otherwise we don't have something we can lock.

Reply:
Never mind. In hindsight it's not worth it.